tests/test_unit/test_cdf_tk/test_utils/test_safe_dict.py

```diff
@@ -0,0 +1,283 @@
+        assert len(d) == num_threads * operations_per_thread
+        for thread_id in range(num_threads):
+            for i in range(operations_per_thread):
+                key = thread_id * 1000 + i
+                assert d[key] == key * 3
+
+    def test_concurrent_modifications(self):
+        """Test concurrent modifications don't cause corruption."""
+        d = ThreadSafeDict[str, int]()
+        num_threads = 20
+
+        def worker(thread_id: int):
+            for i in range(50):
+                key = f"thread_{thread_id}_key_{i}"
+                d[key] = thread_id * 100 + i
+
+                # Sometimes delete keys from other threads
+                if i % 10 == 0:
+                    for other_thread in range(thread_id):
+                        other_key = f"thread_{other_thread}_key_{i}"
+                        d.pop(other_key, None)
+
+        with ThreadPoolExecutor(max_workers=num_threads) as executor:
+            futures = [executor.submit(worker, i) for i in range(num_threads)]
+            for future in as_completed(futures):
+                future.result()  # Wait for completion and check for exceptions
+
+        # Verify the dict is still consistent (no corruption)
+        # Count remaining keys
+        remaining_keys = len(d)
+        assert remaining_keys >= 0  # Should not be negative due to corruption
+
+    def test_exception_handling(self):
+        """Test proper exception handling."""
+        d = ThreadSafeDict[str, int]()
+
+        # KeyError on missing key
+        with pytest.raises(KeyError):
+            _ = d["nonexistent"]
+
+        # KeyError on pop of missing key
+        with pytest.raises(KeyError):
+            d.pop("nonexistent")
+
+        # KeyError on popitem from empty dict
+        with pytest.raises(KeyError):
+            d.popitem()
+
+        # KeyError on del of missing key
+        with pytest.raises(KeyError):
+            del d["nonexistent"]
+
+    def test_pop_with_default(self):
+        """Test pop method with default value."""
+        d = ThreadSafeDict({"a": 1})
+
+        # Pop existing key
+        assert d.pop("a") == 1
+        assert len(d) == 0
+
+        # Pop non-existing key with default
+        assert d.pop("b", 42) == 42
+
+        # Pop non-existing key without default should raise KeyError
+        with pytest.raises(KeyError):
+            d.pop("c")
+
+    def test_update_methods(self):
+        """Test various ways to update the dictionary."""
+        d = ThreadSafeDict[str, int]()
+
+        # Update with dict
+        d.update({"a": 1, "b": 2})
+        assert d["a"] == 1
+        assert d["b"] == 2
+
+        # Update with keyword arguments
+        d.update(c=3, d=4)
+        assert d["c"] == 3
+        assert d["d"] == 4
+
+        # Update with iterable of pairs
+        d.update([("e", 5), ("f", 6)])
+        assert d["e"] == 5
+        assert d["f"] == 6
+
+    def test_performance_under_contention(self):
```

Comment:
Will this test or other tests run for many seconds or more?  
That could slow down CI? 
Maybe some of them should be skipped by default, only enabled for manual runs?

Reply:
4 ms is an acceptable runtime for the test :)